ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -25,57 +25,951 @@
+			return true;
+		}
+		break;
+	case LT_EQ:
+		if (0 == strncmp(_cursor, "<=", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case GT_GT:
+		if (0 == strncmp(_cursor, ">>", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case LT_LT:
+		if (0 == strncmp(_cursor, "<<", 2)) {
+			_cursor += 2;
+			return true;
+		}
+		break;
+	case PLUS:
+		if (('+' == _cursor[0]) && ('+' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case MINUS:
+		if (('-' == _cursor[0]) && ('-' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case MULT:
+		if ('*' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case DIV:
+		if ('/' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case REM:
+		if ('%' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case NOT:
+		if (('!' == _cursor[0]) && ('=' != _cursor[1])) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case TILDE:
+		if ('~' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case LPAREN:
+		if ('(' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	case RPAREN:
+		if (')' == _cursor[0]) {
+			_cursor += 1;
+			return true;
+		}
+		break;
+	}
+	return false;
+}
+
+/**
+ * Reads an integer number starting at the cursor in the expression string.
+ * Number is read as a 64bit integer.
+ */
+bool
+MacroScanner::readNumber(int64_t *ret)
+{
+	uint64_t retval = 0;
+	int base = 0;
+	bool negative = false;
+
+	while (isspace(_cursor[0])) {
+		_cursor += 1;
+	}
+
+	if ((_cursor < _end) && ('+' == _cursor[0] || '-' == _cursor[0])) {
+		negative = '-' == _cursor[0] ;
+		_cursor += 1;
+	}
+
+	if ((_cursor < _end) && !(isdigit(_cursor[0]))) {
+		return false;
+	}
+
+	if ((_cursor < _end) && ('0' == _cursor[0])) {
+		_cursor += 1;
+		if (('b' == _cursor[0]) || ('B' == _cursor[0])) {
+			base = 2;
+			_cursor += 1;
+		} else if (('x' == _cursor[0])  || ('X' == _cursor[0])) {
+			base = 16;
+			_cursor += 1;
 		} else {
-			/* For macros containing brackets, extract the number. This works for casts
-			 * such as ((int)5).
-			 */
-			rc = DDR_RC_ERROR;
-			while (string::npos != value.find('(')) {
-				size_t lastOpenParen = value.find_last_of('(');
-				size_t nextCloseParen = value.find(')', lastOpenParen);
-				if (string::npos == nextCloseParen) {
-					break;
+			base = 8;
+		}
+	} else {
+		base = 10;
+	}
+
+	while ((_cursor < _end) && isadigit(_cursor[0], base)) {
+		retval *= base;
+		retval += digitvalue(_cursor[0]);
+		_cursor += 1;
+	}
+
+	/* type suffixes do nothing since we treat each literal as an uint64_t.
+	 * so we just verify the suffix is correct
+	 */
+	int numL = 0;
+	int numU = 0;
+	for (int i= 0; (_cursor < _end) && (i < 3); i++, _cursor++) {
+		if (('l' == _cursor[0]) || ('L' == _cursor[0]) ) {
+			numL += 1;
+		} else if (('u' == _cursor[0]) || ('U' == _cursor[0])){
+			numU += 1;
+		} else {
+			break;
+		}
+	}
+	if (2 < numL || 1 < numU) {
```

Comment:
Parentheses missing here.
I personally find it difficult to read such comparisons: I think `((numL > 2) || (numU > 1))` is clearer.